File: src/types/cipher.ts
import { Fpt, Fr, G1, G1Projective, G2, G2Projective, Gt } from "./elements.ts"

/**
 * The `Cipher` interface represents a ciphersuite, which is a set of cryptographic algorithms used to implement a
 * specific signature scheme. This interface defines terms specific to a pairing-friendly elliptic curve.
 */
export interface Cipher {
  /**
   * The unique identifier for the ciphersuite, which will be represented as an ASCII encoded octet string.
   *
   * The REQUIRED format for this string is `BBS_ || H2C_SUITE_ID || ADD_INFO`, where:
   *  - `H2C_SUITE_ID` is the identifier of the hash-to-curve suite used by the ciphersuite, and
   *  - `ADD_INFO` is an optional additional octet string indicating any additional information used to uniquely qualify
   *    the ciphersuite. When specified, this value MUST only contain ASCII encoded characters with codes between 0x21
   *    and 0x7E, inclusive, and MUST end with an underscore.
   */
  id: Uint8Array

  /**
   * The cryptographic hash function used by the ciphersuite.
   */
  hash: string

  /**
   * Number of bytes to represent a scalar value, in the multiplicative group of integers mod r, encoded as an octet
   * string.
   *
   * It is RECOMMENDED that this value be set to `ceil(log2(r)/8)`.
   */
  octetScalarLength: number

  /**
   * Number of bytes to represent a point, encoded as an octet string.
   */
  octetPointLength: number

  /**
   * The operation used to expand a message to a scalar value.
   *
   * This value MUST be defined to be at least `ceil(log2(r)+k/8)`, where `r` and `k` are defined by the ciphersuite.
   */
  expandLength: number

  /**
   * A fixed point in the G1 group, different from the base point `bp1`.
   *
   * This leaves the base point `bp1` free for use in other protocols, like key commitment and proof of possession.
   */
  p1: G1Projective

  /**
   * The fields and groups of the ciphersuite.
   */
  types: {
    G1: G1
    G2: G2
    Fr: Fr
    Fpt: Fpt
  }

  /**
   * A cryptographic hash function that takes an arbitrary octet string as input, and returns a point in G1.
   *
   * @param {Uint8Array} message An octet string representing the message to be hashed.
   * @param {Uint8Array} dst The domain separation tag.
   *
   * @returns {G1Projective} A point in G1.
   */
  hashToCurveG1(message: Uint8Array, dst: Uint8Array): G1Projective

  /**
   * The operation used to expand a message to a scalar value.
   *
   * @param {Uint8Array} message An octet string representing the message to be expanded.
   * @param {Uint8Array} dst The domain separation tag.
   * @param {number} expandLength The number of bytes to expand the message to.
   *
   * @returns {Uint8Array} An octet string representing the expanded message.
   */
  expandMessage(message: Uint8Array, dst: Uint8Array, expandLength?: number): Uint8Array

  /**
   * The non-degenerate bilinear pairing operation, mapping a point in G1 and and a point in G2 to a point in Gt.
   *
   * @param {G1Projective} point1 A point in G1.
   * @param {G2Projective} point2 A point in G2.
   *
   * @returns {Gt} The result of the pairing operation.
   */
  pairing(point1: G1Projective, point2: G2Projective): Gt

  /**
   * Compare two pairings for equality.
   *
   * @param {Gt} pair1 The first pairing to compare.
   * @param {Gt} pair2 The second pairing to compare.
   * @param {Gt} pair3 The third pairing to compare.
   */
  pairingCompare(pair1: Gt, pair2: Gt, pair3: Gt): boolean
}

File: src/suite/cipher.rs
use crate::suite::constants::LENGTH_G1_POINT;
use bls12_381::{G1Affine, G2Prepared, Gt};

pub struct Cipher {
    /// The unique identifier for the cipher suite, which will be represented as an ASCII string.
    ///
    /// The REQUIRED format for the string is `BBS_ || <h2c_suite_id> || <add_info>`, where:
    /// - `BBS_` is a constant prefix.
    /// - `<h2c_suite_id>` is the identifier of the hash-to-curve suite, and
    /// - `<add_info>` is an optional string that indicating any additional information used to uniquely identify the
    ///         suite. When specified, this value MUST only contain ASCII characters with codes between 0x21 and 0x7E,
    ///         inclusive, and MUST end with an underscore.
    pub id: &'static [u8],

    /// A fixed point in the G1 group, different from the identity element.
    ///
    /// This leaves the identity element free for use in other protocols, like key commitment and proof of possession.
    pub singularity: [u8; LENGTH_G1_POINT],

    /// A cryptographic hash function that takes an arbitrary octet string as input, and returns a point in G1.
    ///
    /// - `msg`: An octet string representing the message to be hashed.
    /// - `dst`: A domain separation tag.
    ///
    /// Return a point in G1.
    pub hash_to_curve: fn(message: &[u8], dst: &[u8]) -> G1Affine,

    /// The operation used to expand a message to a scalar.
    ///
    /// - `msg`: An octet string representing the message to be expanded.
    /// - `dst`: A domain separation tag.
    /// - `len`: The number of bytes to expand the message to.
    ///
    /// Return an octet string representing the expanded message.
    pub expand_message: fn(message: &[u8], dst: &[u8], expand_length: Option<usize>) -> Vec<u8>,

    /// Compare two pairing results for equality.
    ///
    /// - `terms`: A list of pairs of points in G1 and G2.
    /// - `result`: The target pairing result, a point in Gt.
    ///
    /// Return `true` if the product of the pairing results are equal to the target result.
    pub pairing_compare: fn(terms: &[(&G1Affine, &G2Prepared)], result: &Gt) -> bool,
}
